compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -3766,69 +3806,100 @@ abstract class GenJSCode extends plugins.PluginComponent
+                  retTpe)
             }
           }, { // else
             callStatement
           })(jstpe.AnyType)
         }
+
+        /* For primitive types, we need to handle two cases. The method could
+         * either be defined in the boxed class of the primitive type, or it
+         * could be defined in the primitive class itself. In the former case,
+         * we generate a direct call to the actual (hijacked) method. In the
+         * latter case, we directly generate the primitive operation.
+         */
+
+        def addCallForPrimitive(primitiveClass: Symbol): Boolean = {
+          val boxedClass = definitions.boxedClass(primitiveClass)
+          val boxedTpe = boxedClass.tpe
+          val methodInBoxedClass = matchingSymIn(boxedClass)
+          if (methodInBoxedClass != NoSymbol && methodInBoxedClass.isPublic) {
+            callStatement = js.If(genIsInstanceOf(callTrg, boxedTpe), {
+              boxIfNeeded(
+                  genApplyMethod(
+                      genAsInstanceOf(callTrg, boxedTpe),
+                      methodInBoxedClass,
+                      arguments),
+                  methodInBoxedClass.tpe.resultType)
+            }, { // else
+              callStatement
+            })(jstpe.AnyType)
+            true
+          } else {
+            val methodInPrimClass = matchingSymIn(primitiveClass)
+            if (methodInPrimClass != NoSymbol && methodInPrimClass.isPublic) {
+              def isIntOrLongKind(kind: TypeKind) = kind match {
+                case CharKind     => false
+                case _:INT | LONG => true
+                case _            => false
+              }
+              val ignoreBecauseItMustBeAnInt = {
+                primitiveClass == DoubleClass &&
+                toTypeKind(methodInPrimClass.tpe.resultType) == DoubleKind &&
+                isIntOrLongKind(toTypeKind(sym.tpe.resultType))
+              }
+              if (ignoreBecauseItMustBeAnInt) {
+                // Fall through to the Int case that will come next
+                false
+              } else {
+                callStatement = js.If(genIsInstanceOf(callTrg, boxedTpe), {
+                  val castCallTrg =
+                    makePrimitiveUnbox(callTrg, primitiveClass.tpe)
+                  val call = genPrimitiveOpForReflectiveCall(methodInPrimClass,
+                      castCallTrg, arguments)
+                  boxIfNeeded(call, methodInPrimClass.tpe.resultType)
+                }, { // else
+                  callStatement
+                })(jstpe.AnyType)
+                true
+              }
+            } else {
+              false
+            }
+          }
+        }
+
+        addCallForPrimitive(BooleanClass)
+        addCallForPrimitive(LongClass)
+        addCallForPrimitive(CharClass)
+
+        /* For primitive numeric types that box as JS numbers, find the first
+         * one that matches. It will be able to handle the subsequent cases.
```

Comment:
This matching logic is very hard to understand and verify it is correct. Have you considered handling the sets of relevant methods individually? There are not that many:

- Conversions `toX`
- Unary ops
- String concat
- Comparisons
- Binary number ops
- Binary int ops

I'm not sure that it is feasible. Maybe an acceptable compromise is to order the classes according to the runtime subtyping relationship and generate all cases. This would get rid of the strange `ignoreBecauseItMustBeAnInt` and probably be easier to understand :-/